File: 872-leaf-similar-trees.py
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def dfs(root,ans):
            if not root:
                return
            if not root.left and not root.right:
                ans.append(root.val)
            dfs(root.left,ans)
            dfs(root.right,ans)
            return ans
        ans1 = []
        ans2 = []
        return dfs(root1,ans1) == dfs(root2,ans2)

File: 872-leaf-similar-trees.java
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> leaves1 = new ArrayList<>();
        List<Integer> leaves2 = new ArrayList<>();
        
        dfs(root1, leaves1);
        dfs(root2, leaves2);
        
        return leaves1.equals(leaves2);
    }
    private void dfs(TreeNode root, List<Integer> res){
        if(root == null) return ;
        if(root.left == null && root.right == null) res.add(root.val);
        dfs(root.left,res);
        dfs(root.right,res);
        return;
    }
}
